Fleet Management Integration, Teltonika Telematics Device Support

pankalog opened the pull request:
This is still missing tests, which I plan to implement this week. 

The UI folder will be removed from the PR before merging, it is just useful for debugging purposes.

pankalog committed:
Reads Teltonika JSON payloads
and either generates or updates an asset
depending on the IMEI of the device.

pankalog committed:
Reads Teltonika JSON payloads
and either generates or updates an asset
depending on the IMEI of the device.

Currently reads Teltonika payloads,
finds the Asset using the IMEI, and
determines whether or not
a new Asset needs to be created
or an old one needs to be updated.

pankalog committed:
Unsecured auto-provisioning.
Creates asset when new IMEI is used,
Updates the asset if the IMEI has already
been used.

pankalog committed:
New method for parsing Attributes from payload

The method will parse the payload
coming from the device
to decode it into an attribute list.

pankalog committed:
New method for parsing Attributes from payload

The method will parse the payload
coming from the device
to decode it into an attribute list.

pankalog committed:
First implementation of the Car Asset

This asset helps with conciseness on code in the MQTT Handler, since we are aware that the required attributes exist and are complete.

pankalog committed:
MQTT Handler V1.

Parses the JSON payload according to the screenshot on Teltonika forums, links the parameter IDs to their descriptions based on the Teltonika wiki, and dynamically inserts or updates an asset with the most recent data.

TODO:
- Proper Type definition for each attribute depending on the `Type` parameter in the Teltonika parameter ID forums

- Secure communication and provisioning of Teltonika Devices

- Automatically retrieve Teltonika device parameter information (preferrably from an official API)

pankalog committed:
Location History Overlay V1

When an asset marker is clicked, its history from midnight to the current time is displayed as lines and points. When deselected, the GeoJSON layers and sources are removed from the MapLibre map object.

TODO: Improve overall type-safety and robustness, improve UI line and point elements, implement configuration item to specifically allow the overlay to be visible.

pankalog committed:
Location History Overlay V1.01

Overlay is properly removed when another asset marker is selected, and when no asset is selected.

pankalog committed:
Tester commit

pankalog committed:
Revert map chart changes

pankalog committed:
MQTT Handler Improvements

- Now does not require authentication for publishing to Teltonika topic (while Authentication is being worked on)

- Improved robustness and added more logging for when devices connect and publish.

pankalog committed:
Teltonika MQTT Handler Improvements

- Since I have a device in my hands, I can test out the payload, and adjust the jackson parser.

pankalog committed:
Update .env

pankalog committed:
Update docker-compose.yml

pankalog committed:
Teltonika MQTT Handler Improvements

- FMC003.json is not being included in the build, quick and dirty fix to get the parameters to be read accordingly.

pankalog committed:
Teltonika MQTT Handler Improvements

- include FMC003.json in the build, quick and dirty fix to get the parameters to be read accordingly.

pankalog committed:
Teltonika MQTT Handler Improvements

- Directory fixes

pankalog committed:
Teltonika MQTT Handler Improvements

- Directory fixes and change string

pankalog committed:
Teltonika MQTT Handler Improvements

- Fixed bug that only read the first attribute;
- Now logs the amount of Teltonika AVL parameters from file

pankalog committed:
Teltonika MQTT Handler Improvements

- Fixed latitude and longitude regex to properly recognize all coordinates

pankalog committed:
Teltonika MQTT Handler Improvements

- Fixed latitude and longitude regex to properly recognize all coordinates
- Select the Europe MBTILES file for deployment

pankalog committed:
Teltonika MQTT Handler Improvements

- Correctly parse the coordinates sent by the device

pankalog committed:
Parameter Type Parser WIP

pankalog committed:
TeltonikaMQTTHandler V2, some inactive Maps features.

- MQTT Handler updated to correctly parse data from real devices
- State.java now correctly parses the Type of each parameter
- State.java now correctly includes Attribute Metadata, based on TeltonikaParameter
- MQTT Handler now correctly handles updating and creating an Asset's Attributes using SEDA messages

pankalog committed:
Payload State Parser Bug Fixes

- Bug when trying to render a Unit of Measurement that is not defined in the Constants file

pankalog committed:
Updated docker-compose.yml to use mbtiles file for Europe

pankalog committed:
TeltonikaMQTTHandler V2.5

- Support for sending messages to the Teltonika devices.
- Bug fixes and refactors

pankalog committed:
TeltonikaMQTTHandler V2.5

- Formatted the command to be of the appropriate format for the device

pankalog committed:
TeltonikaMQTTHandler V2.5

- Bug fixes

pankalog committed:
TeltonikaMQTTHandler V2.5

- Bug fixes
- MQTT Handler now parses TeltonikaMessageResponses, the response to messages sent to the device, and stores them to the "response" attribute.

pankalog committed:
Deployment fix

pankalog committed:
Documentation, Refactors, Bug-fixes, improvements

pankalog committed:
AssetStateDuration ValueType and TeltonikaMQTTHandler use-case.

pankalog committed:
Fleet Management v3: Trip (or state/session) Recognition and recording

mapwidget.ts: Implemented reverse geocode support
or-map-asset-card.ts: Added or-map-asset-card-trip-section.ts
or-map-asset-card-trip-section.ts: Prints relevant trip data per-trip
State.java: Bug-fixes, improvements, timestamp recognition re-write
TeltonikaMQTTHandler.java: Bug-fixes, improvements:
Implemented TeltonikaMQTTHandler.assetChangedTripState, along with AssetStateDuration. Wrote driver-code to recognize when a trip is started using AttributePredicate.

pankalog committed:
Update TeltonikaMQTTHandler.java

Requires a backslash to correctly refer to Teltonika Parameter file

pankalog committed:
Added exception catches for debugging

pankalog committed:
Merge pull request #1 from pankalog/feature/fleet-management

Feature/fleet management

pankalog committed:
Remove UI Elements

pankalog committed:
Remove UI Elements

pankalog committed:
Code Cleanup for PR #1147

pankalog committed:
Merge pull request #1148 from pankalog/master

Merge from fork

pankalog committed:
Code Cleanup 2

pankalog commented:
There are various notices that need to be given to all that peer into this pull request;

First, there is no tests here because I am not sure what the testing strategy is. Once I recognize exactly what is expected I will of course follow up with a proper testing strategy and structure. 

Second, the current state of this PR is not going to be mergeable for the above reasons. I am not expecting an all-clear, but I am expecting comments about what has already been done, and what needs to yet be done. 

OpenRemote is a very vast project with a lot of design choices and I am sure I have not made every right choice. I feel that a lot of code might be misplaced, re-written, or incorrectly used. I would like to mostly focus on issues like that.

I am also not sure if FMC003.json should or should not be included within the OpenRemote package. 

Can't wait for your comments!

pankalog committed:
Security fixes and initial batch of tests

pankalog committed:
New IMEIValidator class and bug fixes in ContainerTrait

pankalog committed:
Merge pull request #1152 from pankalog/master

Tests merge

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+    // To be removed when auto-provisioning works
+    @Override
+    public boolean canSubscribe(RemotingConnection connection, KeycloakSecurityContext securityContext, Topic topic) {
+        if(topic.getTokens().size() < 5){
```
In other MQTT implementations I see the use of `topicMatches()` override.
Maybe worth practicing here too?

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+    @Override
+    public void onPublish(RemotingConnection connection, Topic topic, ByteBuf body) {
+        String payloadContent = body.toString(StandardCharsets.UTF_8);
+        getLogger().info(payloadContent);
```
Remove these logging statements before merging.

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+    }
+
+    @Override
+    public void onPublish(RemotingConnection connection, Topic topic, ByteBuf body) {
```
Please add some comments / context to this method.
I'm not fully known with the use case, but for me it's weird to create assets during `onPublish()` ?
Especially since no actual payload is published?

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+     * @param realm The realm to create the Asset in.
+     * @param attributes The attributes to insert in the Asset.
+     */
+    private void CreateNewAsset(String newDeviceId, String newDeviceImei, String realm, AttributeMap attributes) {
```
We practice lowerCamelCase methods ;)

pankalog replied:
Fixed everywhere in TeltonikaMQTTHandler

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+        Asset<?> mergedAsset = assetStorageService.merge(testAsset);
+
+        if(mergedAsset != null){
+            getLogger().info("Created Asset through Teltonika: " + testAsset);
```
Be hesitant when adding log statements. They shouldn't be overdone let's say

pankalog replied:
Removed a lot of logs from TeltonikaMQTTHandler

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+                params.put(param.getPropertyId(), param);
+            }
+        } catch (Exception e) {
+            getLogger().info(e.toString());
```
`info()` instead of `severe()` ?

pankalog replied:
Added better verbosity

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+            // Add each element to the HashMap, with the key being the unique parameter ID and the parameter
+            // being the value
+            for (TeltonikaParameter param : paramArray) {
+                params.put(param.getPropertyId(), param);
```
There is probably an more efficient / elegant solution to this. Java streams?

pankalog replied:
Implemented using Java streams

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+
+            mapper = new ObjectMapper();
+            TeltonikaResponsePayload response = mapper.readValue(payloadContent, TeltonikaResponsePayload.class);
+            getLogger().info(response.rsp);
```
Remove this before merging.

pankalog replied:
Removed

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+        //before the newValue.
+        //This means that, the state change took place between the datapoint we just found and its next one.
+
+        //Find the first datapoint that passes the negated predicate
```
Just as a practice; I would be hesitant in creating lots of multiline comments. Just for readability you know haha

pankalog replied:
Maximum of 2 lines per comment, from now on, I hope nobody minds!

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+
+        //Find the first datapoint that passes the negated predicate
+
+        AssetDatapoint StateChangeAssetDatapoint = null;
```
We practice lowerCamelCase variables ;)

pankalog replied:
Renamed all functions

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+                    throw new Exception("Found state change datapoint failed predicate");
+                }
+            }else{
+                throw new Exception("Couldn't find asset state change value");
```
I know it's being catched by line 529, but throwing a more specific exception type helps readability / future cases.

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+
+        //BUG: The Timestamp of an asset datapoint is NOT attribute.getTimestamp()!
+        //This makes it much harder to relate datapoints to relate a duration to the time-distance between two datapoints.
+        //To make this easier, we are going to use some time-padding of 1 second from the gap. It's going to be sufficient to cover some edge-cases.
```
Hmmmm interesting finding! Don't know if we can change this in the datapoint code.
You can create a GitHub issue for this. I guess you can also assign it to me.

Don't know how large the gap is between an datapoint timestamp and an attributeChanged timestamp?

I would hesitate to practice a "margin of error" of 1 second.
Imagine the system being slow, taking for example 500+ ms to handle an attribute change, it could hurt this.
Maybe something like querying the datapoint with the closest timestamp and the same value would work?

pankalog replied:
Of course the margin of error was added at least so that I can get my data in, because without it I lost a lot of datapoints. It's not optimal in any way. 

Lets continue this in #1156.

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+                new Timestamp(previousValue.getTimestamp().get())
+        ));
+
+        tripAttr.setTimestamp(StateChangeAssetDatapoint.getTimestamp());
```
... I guess this works too. (see previous comment)

pankalog replied:
Still does not solve the issue, it's not a great way to solve this because I also get some "false" trips

MartinaeyNL commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+        String imei = asset.getAttribute("IMEI").toString();
+
+        getLogger().info("Updating CarAsset with IMEI "+imei);
+        //DONE: If the parameters do not exist, then the update fails,
```
Remove these notes when merging I guess? Don't know how important they are?

It's also a kind of general practice to not use too many multiline comments in methods.
Placing a large comment above the function describing it is indeed a good habit,
but in the method itself it should only shortly describe its meaning and possible behavior.

pankalog replied:
Removed a lot of them, but I am keeping the TODOs until I have a very stable (or merge-able) version. When I do, this will be converted to proper entries in a Kanban board

MartinaeyNL commented on test/src/test/groovy/org/openremote/test/protocol/mqtt/TeltonikaMQTTProtocolTest.groovy:
```diff
@@ -0,0 +1,284 @@
+        client = new MQTT_IOClient(mqttClientId, mqttHost, mqttPort, false, false, null, null, null)
+    }
+
+    def cleanupSpec(){
```
I think cleanup methods shouldn't be necessary. See `MqttBrokerTest` for inspiration.

pankalog replied:
I think they are because I use multiple tests for my groovy tests, and they just help with keeping things neat I believe.

MartinaeyNL commented:
For context;

@pankalog asked me for a review of his draft.
I myself don't really have much knowledge of the MQTTHandler, neither much of MQTT (other than the basics)
Will need additional review when it's in a ready state.

MartinaeyNL commented:
To respond on your comment @pankalog 

In general I believe architecture-wise you did a great job in implementing your own MQTTHandler.
And you did a good effort in providing tests to cover your additional additions.

Of course I can't fully agree nor disagree with architecture things, since @richturner is more experienced with the backend architecture than I am. But in general it looks good!

Let me know if you need anything 😉

pankalog committed:
Teltonika Parameter file and Teltonika payload file that is used for the tests

pankalog committed:
Added tests to cover parameter parsing and bidirectional messaging between client and broker

pankalog committed:
Bug fix

pankalog committed:
Bug fixes

pankalog committed:
Bug fixes

pankalog committed:
Merge branch 'master' into feature/fleet-management

pankalog committed:
Bug fixes and reverting changes from ContainerTrait. Fixed TeltonikaMQTTHandler.canSubscribe.

pankalog committed:
Bug fixes

pankalog committed:
Merge branch 'feature/fleet-management' into master

pankalog committed:
Bug fixes

pankalog committed:
Bug fixes

pankalog committed:
Tests for correct payload parsing and bidirectional communication, bug-fixes

Tests for correct payload parsing and bidirectional communication, bug-fixes

pankalog committed:
Merge pull request #2 from openremote/feature/fleet-management

Tests for correct payload parsing and bidirectional communication, bu…

pankalog commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+        if (!Objects.equals(event.getAttributeName(), DEVICE_SEND_COMMAND_ATTRIBUTE_NAME)) return;
+        getLogger().info(event.getEventType());
+        //Find the asset in question
+        Asset<?> asset = assetStorageService.find(new AssetQuery().ids(event.getAssetId()).types(CarAsset.class));
```
That's a great method, I need to refactor everything to use it in a lot of places now

pankalog committed:
Bug fixes and fix all problems

pankalog committed:
Merge pull request #1157 from pankalog/master

Bug fixes

pankalog commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+     */
+    @Override
+    public boolean checkCanPublish(RemotingConnection connection, KeycloakSecurityContext securityContext, Topic topic) {
+        return true;
```
The reason I'm currently overriding this is because of avoiding Keycloak authentication, something currently not possible for Teltonika devices. This will be removed when authentication for the devices is properly implemented

pankalog commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,639 @@
+    @Override
+    public boolean canPublish(RemotingConnection connection, KeycloakSecurityContext securityContext, Topic topic) {
+        getLogger().info("Teltonika device will publish to Topic "+topic.toString()+" to transmit payload");
+        return true;
```
Removed all Keycloak usages, especially in TeltonikaMQTTHandler#start, since Keycloak is currently not required for any function of this Handler.

pankalog committed:
Bug fixes, conciseness, whitespaces

pankalog commented on model/src/main/java/org/openremote/model/asset/impl/CityAsset.java:
Reverted

pankalog commented on model/src/main/java/org/openremote/model/teltonika/IMEIValidator.java:
I couldn't find a validator for IMEIs within OpenRemote, and I wanted to use a fully-implemented validator with checksums.

pankalog commented on test/src/testFixtures/groovy/org/openremote/test/ContainerTrait.groovy:
Reverted

pankalog commented on model/src/main/java/org/openremote/model/value/ValueType.java:
```diff
@@ -130,6 +131,8 @@ public static class MultivaluedStringMap extends HashMap<String, List<String>> {
     );
 
+    public static final ValueDescriptor<AssetStateDuration> ASSET_STATE_DURATION = new ValueDescriptor<>("AssetStateDuration", AssetStateDuration.class);
+
```
We can talk about renaming at #1146

pankalog commented:
I just went through @MartinaeyNL 's review comments and I resolved as many as I can. The rest of them cannot really be resolved right now without further approval/definition of what needs to do. Waiting for further reviews!

pankalog committed:
Merge pull request #3 from openremote/feature/fleet-management

Feature/fleet management

pankalog committed:
Merge pull request #4 from openremote/master

Merge master from openremote/openremote

pankalog committed:
Fixed constraint meta items for Teltonika parameters, fixed issues after merge from master

pankalog committed:
Merge pull request #5 from openremote/master

Console settings improvements (#1143)

pankalog committed:
Merge pull request #1160 from pankalog/master

Merge bug fixes from fork

pankalog committed:
Fixed merge issues for tests

pankalog committed:
Tried to fix Timestamping issues

pankalog committed:
Merge pull request #1167 from pankalog/master

Tried to fix Timestamping issues

pankalog committed:
Merge pull request #6 from openremote/master

merge from openremote/master

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,627 @@
+public class TeltonikaMQTTHandler extends MQTTHandler {
+
+    //TODO: Maybe move this to Models?
+    private static class TeltonikaDevice {
```
Prefer `protected` as it easily allows others to subclass your code and make changes.

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,627 @@
+        assetStorageService = container.getService(AssetStorageService.class);
+        AssetDatapointService = container.getService(AssetDatapointService.class);
+        timerService = container.getService(TimerService.class);
+        DeviceParameterPath = Paths.get("deployment/manager/fleet/FMC003.json");
```
This could be baked into the resources of the module unless a user needs to change it.

This relative path won't work when image is packaged into docker.

pankalog replied:
Can you show me an example for this so that I can implement it?

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,627 @@
+
+        clientEventService.addInternalSubscription(
+                AttributeEvent.class,
+                buildAssetFilter(),
```
`AssetFilter` is a bit limited in its' capability I would suggest using `null` and then filtering the incoming `AttributeEvents` for asset type car etc.

pankalog replied:
Could I still use the buildAssetFilter() method's returned AssetFilter using apply() to check if the event applies? This way we retain the same filter functionality but dynamically refresh the applicable asset list on every message.

pankalog replied:
Here is what I am trying to do when entering handleAttributeMessage(): 

`AssetFilter<AttributeEvent> eventFilter = buildAssetFilter();

        if(eventFilter.apply(event) != null) {
            getLogger().finer("eventFilter applied and the event is not for me");
            return;
        }`

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,627 @@
+
+    private void handleAttributeMessage(AttributeEvent event) {
+        // If this is not an AttributeEvent that updates a TELTONIKA_DEVICE_SEND_COMMAND_ATTRIBUTE_NAME field, ignore
+        if (!Objects.equals(event.getAttributeName(), TELTONIKA_DEVICE_SEND_COMMAND_ATTRIBUTE_NAME)) return;
```
Check asset type of event also

pankalog replied:
https://github.com/openremote/openremote/pull/1147#discussion_r1401988178

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,627 @@
+            if(asset.getAttributes().containsKey(attribute.getName())){
+                AttributeEvent attributeEvent = new AttributeEvent(asset.getId(), attribute.getName(), attribute.getValue(), packetTimestamp);
+                LOG.finer("Publishing to client inbound queue: " + attributeEvent);
+                messageBrokerService.getFluentProducerTemplate()
```
You can use `AssetProcessingService.sendAttributeEvent()`

pankalog replied:
Should I use `AttributeEvent.Source#SENSOR` as the `Source` in the overload of that function? Since it sort of is a sensor update, just handled by the MQTT Handler.

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,627 @@
+     */
+    private void sendCommandToTeltonikaDevice(String command, TeltonikaDevice device) {
+        HashMap<String, String> cmdtest = new HashMap<>();
+        cmdtest.put("CMD", command);
```
`Map.of` is a convenient shorthand

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,627 @@
+
+    @Override
+    public boolean canPublish(RemotingConnection connection, KeycloakSecurityContext securityContext, Topic topic) {
+        getLogger().info("Teltonika device will publish to Topic "+topic.toString()+" to transmit payload");
```
Log level should be lower as this will be hit on every device publish

pankalog replied:
set to `FINER`

richturner commented on model/src/main/java/org/openremote/model/asset/impl/CarAsset.java:
```diff
@@ -0,0 +1,46 @@
+    public static final AttributeDescriptor<String> MAKE_AND_MODEL = new AttributeDescriptor<>("makeAndModel", ValueType.TEXT).withOptional(true);
+    public static final AttributeDescriptor<Integer> MODEL_YEAR = new AttributeDescriptor<>("modelYear", ValueType.INTEGER).withOptional(true);
+    /**
+     * If Rich sees this, sorry for the American spelling!
```
That's it no more assistance from me

richturner commented on model/src/main/java/org/openremote/model/teltonika/IMEIValidator.java:
```diff
@@ -0,0 +1,48 @@
+ * @see <a href="https://www.gsma.com/newsroom/wp-content/uploads//TS.06-v16.0.pdf">the GSMA IMEI Allocation Guidelines</a>.
+ *
+ */
+public class IMEIValidator {
```
A protected constructor will prevent instances of this helper class

richturner commented on model/src/main/java/org/openremote/model/teltonika/PayloadJsonObject.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Override
+    public String toString() {
```
Is this for logging purposes?

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+import java.util.regex.Pattern;
+
+import static org.openremote.model.value.MetaItemType.*;
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
This is default for our Jackson `ObjectMapper`

richturner commented on model/src/main/java/org/openremote/model/teltonika/PayloadJsonObject.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.Map;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({
```
key ordering in JSON is not guaranteed plus you only have one property being serialised

richturner commented on model/src/main/java/org/openremote/model/teltonika/PayloadJsonObject.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
This is default behaviour for our Jackson `ObjectMapper`

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+
+
+    @Override
+    public String toString() {
```
Is this just for use in logging?

pankalog replied:
Yes, should I remove it?

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+    //TODO: Move any function from here; it's a POJO.
+    //TODO: Figure out what the parameters: pr, alt, ang, sat, sp, and evt do
+    // and implement their functionality (I can guess but I want to know exactly).
+    public AttributeMap GetAttributes(Map<Integer, TeltonikaParameter> params, AttributeMap additionalAttributes, Logger logger){
```
camel case is normal in java not pascal case

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+            if (!params.containsKey(parsedEntryId)) continue;
+
+            //Retrieve the parameter data
+            TeltonikaParameter parameter = params.get(Integer.valueOf(entry.getKey()));
```
use `parsedEntryId` here no need to parse the `int` again

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+            }
+            //If the Teltonika Parameter HashMap doesn't contain the requested AVL ID, continue:
+
+            if (!params.containsKey(parsedEntryId)) continue;
+
+            //Retrieve the parameter data
+            TeltonikaParameter parameter = params.get(Integer.valueOf(entry.getKey()));
```
Can just check if `parameter == null`

pankalog replied:
https://github.com/openremote/openremote/pull/1147#discussion_r1402023477

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+                //TODO: Fix frontend to be able to display even custom units, not just the predefined ones
+                //possibly prepend the unit with the string "custom."? So that it matches the predefined format
+                //Add on its units
+                if(!Objects.equals(parameter.units, "-")){
```
`parameter.units` could be null or empty?

pankalog replied:
Since these are the values parsed from the Teltonika wiki, all of the no-values have that dash as a value. Should I do some pre-processing when importing the values to set them to proper values, and then check those values in this method?

pankalog committed:
Added preserveTimestamps to AssetStorageService.merge.

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+
+
+            //Use the AttributeDescriptor and the Value to create a new Attribute
+            Attribute<?> generatedAttribute = new Attribute(attributeDescriptor, value.get());
```
`AttributeDescriptor` are for static use not runtime use; you should just use the other constructor(s):

```
Attribute(String name, ValueDescriptor<T> valueDescriptor, T value)
```

pankalog committed:
Added test for timestamp bug, MQTT Handler changes

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+        //Special parameter definitions without being defined in the AVL Parameter List, thanks Teltonika
+
+        //latlng are the latitude-longitude coordinates, also check if it's 0,0, if it is, don't update.
+        if (reported.containsKey("latlng") && !Objects.equals(reported.get("latlng"), "0.000000,0.000000")){
+            try{
```
unlikely but what if the device is at this actual location

pankalog replied:
The only reason I do this is because the device will send those coordinates if it isn't able to get a satellite fix. Not to mention, this coordinate is quite far out at sea, but at that point, we can think of something else. I am only doing this because I would like the timestamps of attributes to be when the actual last data of the attribute was sent, instead of when dummy data was sent (like 0,0), for other reasons. Also, this is configurable on the device-side, meaning that the user can configure if the device will send packets if there is a satellite fix or not

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+                //Inner method returns Optional.empty, but still throws and prints exception. A lot of clutter, but the exception is handled.
+                value = ValueUtil.getValueCoerced(entry.getValue(), attributeType.getType());
+                if (value.isEmpty()){
+                    attributeType = ValueType.TEXT;
```
This could try to change an existing attributes type which seems odd

pankalog replied:
How do you think this should be handled? Should I not process this attribute or change its type?

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+
+        // Store data points, allow use for rules, and don't allow user parameter modification, for every attribute parsed
+        try{
+            attributes.forEach(attribute -> attribute.addOrReplaceMeta(
```
Maybe this should only be done for new attributes so user can potentially decide to not store data points for an attribute by removing the configuration item

pankalog replied:
Actually, by relating this to https://github.com/openremote/openremote/pull/1147#discussion_r1402049387, should I create some sort of mediary function that checks the existing attribute's attributes against the parsed ones, and then try to merge all of the metadata of the attribute? I have considered this before, but this requires a conversation of what we consider to be more important, the current attributes of the asset or the incoming ones.

pankalog committed:
Merge pull request #1173 from pankalog/master

Merge from fork

richturner commented on manager/src/main/java/org/openremote/manager/asset/AssetStorageService.java:
```diff
@@ -604,17 +604,18 @@ public List<String> findNames(String... ids) {
      */
-    public <T extends Asset<?>> T merge(T asset, boolean overrideVersion) throws IllegalStateException, ConstraintViolationException {
-        return merge(asset, overrideVersion, false, null);
+    public <T extends Asset<?>> T merge(T asset, boolean overrideVersion, boolean preserveTimestamps) throws IllegalStateException, ConstraintViolationException {
```
Don't want to do this unless really necessary would rather solve this problem without a configurable option for consistency; an overload method to maintain the original method signature should have been added

pankalog replied:
This is for-sure not going to remain like this in the future, it is part of the debugging attempts at

richturner commented on manager/src/main/java/org/openremote/manager/system/HealthService.java:
```diff
@@ -63,7 +63,7 @@ public void onInit(Route route) {
     public static final System.Logger LOG = System.getLogger(HealthService.class.getName());
     public static final String OR_METRICS_PORT = "OR_METRICS_PORT";
-    public static final int OR_METRICS_PORT_DEFAULT = 8404;
+    public static final int OR_METRICS_PORT_DEFAULT = 8403;
```
Don't want the default changed here, you can override within your own Run configuration as needed

pankalog replied:
Accidentally committed the file, but for some reason something is already hosted on 8404 at least on my PC, and I'm not exactly sure what

pankalog committed:
Merge pull request #7 from openremote/feature/fleet-management

Feature/fleet management

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+    //TODO: Move any function from here; it's a POJO.
+    //TODO: Figure out what the parameters: pr, alt, ang, sat, sp, and evt do
+    // and implement their functionality (I can guess but I want to know exactly).
+    public AttributeMap GetAttributes(Map<Integer, TeltonikaParameter> params, AttributeMap additionalAttributes, Logger logger){
```
additionalAttributes doesn't seem to be used?

pankalog replied:
It is part of my rethinking of how I should deal with #1156 , meaning if I should pass all attributes I would like to eventually add/update to the Asset or if I should use the returned AttributeMap to append any attributes needed. How should I continue?

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+            try{
+                long unixTimestampMillis = Long.parseLong(reported.get("ts").toString());
+                Timestamp deviceTimestamp = Timestamp.from(Instant.ofEpochMilli(unixTimestampMillis));
+                attributes.add(new Attribute<>("lastContact", ValueType.DATE_AND_TIME, deviceTimestamp).setTimestamp(deviceTimestamp.getTime()));
```
There is an `Attribute` constructor that takes timestamp

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+
+
+
+    private ValueDescriptor<?> GetAttributeType(TeltonikaParameter parameter) {
```
May be better using `ValueUtil.parse()` to convert the string to the correct type and then you can pass the parsed value into here for value type extraction (to prevent parsing the value multiple times which is computationally expensive)

pankalog replied:
I'm not sure I understand, are you talking about checking parameter.type?

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+        }
+    }
+
+    private GeoJSONPoint ParseLatLngToGeoJSONObject(String latlngString) {
```
Make a note of the format of the data being parses

pankalog replied:
Added javadoc

richturner commented on model/src/main/java/org/openremote/model/teltonika/State.java:
```diff
@@ -0,0 +1,298 @@
+    }
+
+    private GeoJSONPoint ParseLatLngToGeoJSONObject(String latlngString) {
+        String regexPattern = "^([-+]?[0-8]?\\d(\\.\\d+)?|90(\\.0+)?),([-+]?(1?[0-7]?[0-9](\\.\\d+)?|180(\\.0+)?))$";
```
Would a simple string split be more effecient here...regex is orders of magnitude slower than a string split

pankalog replied:
I agree, but in this way, we are completely sure that parsing the two numbers is safe and will not throw an exception. Would you want me to convert this to using String.split() and check if the numbers are parsed?

richturner commented on model/src/main/java/org/openremote/model/teltonika/TeltonikaDataPayload.java:
```diff
@@ -0,0 +1,16 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({
```
Same as previous comments

richturner commented on model/src/main/java/org/openremote/model/teltonika/TeltonikaDataPayload.java:
```diff
@@ -0,0 +1,16 @@
+public class TeltonikaDataPayload {
+
+    @JsonProperty("state")
+    public State state;
```
Fields are by default used as json properties so this isn't needed unless you want to change the key name

richturner commented on model/src/main/java/org/openremote/model/teltonika/TeltonikaParameter.java:
```diff
@@ -0,0 +1,141 @@
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({
+        "propertyIdInAvlPacket",
+        "propertyName",
+        "bytes",
+        "type",
+        "min",
+        "max",
+        "multiplier",
+        "units",
+        "description",
+        "hwSupport",
+        "parameterGroup"
+})
+public class TeltonikaParameter {
+
+    @JsonProperty("propertyIdInAvlPacket")
+    public Integer propertyId;
+    @JsonProperty("propertyName")
+    public String propertyName;
+    @JsonProperty("bytes")
+    public String bytes;
+    @JsonProperty("type")
+    public String type;
+    @JsonProperty("min")
+    public String min;
+    @JsonProperty("max")
+    public String max;
+    @JsonProperty("multiplier")
+    public String multiplier;
+    @JsonProperty("units")
+    public String units;
+    @JsonProperty("description")
+    public String description;
+    @JsonProperty("hwSupport")
+    public String hwSupport;
+    @JsonProperty("parameterGroup")
+    public String parameterGroup;
```
Same as other comments

richturner commented on model/src/main/java/org/openremote/model/teltonika/TeltonikaResponsePayload.java:
```diff
@@ -0,0 +1,32 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({
+        "RSP"
+})
```
Same as previous comments

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,654 @@
+
+            if (asset == null) {
+                try{
+                    createNewAsset(deviceUuid, deviceImei, realm, attributes, topic, connection);
```
Should be setting asset to the returned asset from the merge

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,654 @@
+        }
+
+        // Grab all datapoints (To be replaced by AssetDatapointValueQuery)
+        List<AssetDatapoint> valueDatapoints = AssetDatapointService.getDatapoints(ref);
```
Need a limit on this as this could be a large amount of data

pankalog replied:
I have actually been wanting to ask you, this would be much easier and more efficient if I could apply the predicate on values of the database. In fact, I tried implementing a new AssetDatapointQuery that would do that, but I am kinda lost on how that would look like. How do you think I should manage this?

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,654 @@
+        }
+
+        // Grab all datapoints (To be replaced by AssetDatapointValueQuery)
+        List<AssetDatapoint> valueDatapoints = AssetDatapointService.getDatapoints(ref);
```
It would be more efficient to keep in memory the attributes that have an active session rather than searching back through historical data on each publish

pankalog replied:
Would this be stored inside some sort of instance variable in TeltonikaDevice? Hence we could easily link the incoming data/state change to the device producing them. Thanks for the idea!

pankalog replied:
But would it be okay if the state was deleted, in case OpenRemote is shut down? If it was stored in the database, we could simply check the last time the value was equal to the current value.

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,654 @@
+
+            // Add each element to the HashMap, with the key being the unique parameter ID and the parameter
+            // being the value
+            params = Arrays.stream(mapper.readValue(getParameterFileString(), TeltonikaParameter[].class)).collect(Collectors.toMap(
```
You're reading the parameter file in for every payload

pankalog replied:
We had a conversation about how we should manage the parameter files. Would we want to read it once on `TeltonikaMQTTHandler.start()`?

richturner commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,654 @@
+
+        //First merge, then update existing attributes
+        asset.addAttributes(nonExistingAttributes.stream().toArray(Attribute[]::new));
+        if(nonExistingAttributes.size() > 0){
```
isEmpty() method is preferred

pankalog commented on manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java:
```diff
@@ -0,0 +1,654 @@
+     */
+    private void updateAsset(Asset<?> asset, AttributeMap attributes, Topic topic, RemotingConnection connection) {
+
+        Attribute<Date> packetTime = attributes.get(new AttributeDescriptor<Date>("lastContact", ValueType.DATE_AND_TIME)).get();
```
Remnants of a lot of testing runs, there will be no Optional.get() calls without prior checking when ready.

pankalog committed:
Updated test to assure that test asset is deleted before proceeding to next test

pankalog committed:
Revert "Added preserveTimestamps to AssetStorageService.merge."

This reverts commit 59383096ccf91895deba98adc1358e4e0e0f322e.

pankalog committed:
Merge remote-tracking branch 'upstream/feature/fleet-management'

pankalog committed:
AssetStorageService.merge() overload revert, TeltonikaMQTTHandler fix for #1156

pankalog committed:
Merge pull request #8 from pankalog/master

merge from master

pankalog committed:
Merge pull request #1176 from pankalog/feature/fleet-management

merge from fork's Feature/fleet management

pankalog committed:
Private constructor for disallowing instance creation

pankalog committed:
Fixed JSON annotation for Teltonika models

pankalog committed:
Javadoc for GeoJSONPoint parser

pankalog committed:
Big getAttributes rework and optimizations, and reformat

pankalog committed:
Fix method reference after rework

pankalog committed:
Merge remote-tracking branch 'upstream/feature/fleet-management' into feature/fleet-management

pankalog committed:
Merge pull request #1177 from pankalog/feature/fleet-management

Feature/fleet management

pankalog committed:
Bug fixes related to receiving responses to commands from the device, and dealing with timestamps.

pankalog committed:
Comment-out integration test for now.

pankalog committed:
Merge pull request #1178 from pankalog/feature/fleet-management

Feature/fleet management

pankalog committed:
[Test] Do not update the attributes at the updateAsset() method, since they are already set correctly at all other points where attributes are added.

pankalog committed:
Restore map UI elements and AssetStateDuration console-log statements

pankalog committed:
Merge pull request #1179 from pankalog/feature/fleet-management

Feature/fleet management

pankalog committed:
Update AVL ID list directory for deployment support

richturner committed:
Merge remote-tracking branch 'origin/master' into feature/fleet-management

richturner committed:
Initial commit of custom project

richturner committed:
UI update: Fix input field for JSON Object value types (Fixes #1175) (#1184)

richturner committed:
CI/CD: Fix AWS cloud formation template for EC2 host provisioning (#1188)

* CI/CD: Fix cloudformation CFN init ordering
* other minor updates

peterdewitte committed:
Fix purging of syslog events (#1182)

Miggets7 committed:
Console: iOS fix persisting web data (#1197)

Miggets7 committed:
Console: iOS store cookies based on domain (#1198)

Co-authored-by: Michael Rademaker <michael@stofloos.nl>

Miggets7 committed:
Console: iOS missing file for xcode (#1199)

robin-dekkers committed:
Use or-translate for mwc button component (#1206)

wborn committed:
Fix publishing artifacts to Maven Central (#1195)

* Set the publishPassword property to fix authentication issues
* Resolve the proper upload URL for publishing openremote-test

Miggets7 committed:
Consoles: QR provider enhancements and iOS cookie improvements (#1207)

MartinaeyNL committed:
Several small (hot)fixes for reported issues (#1194)

- Fix for Attribute- and Table widget not working in standalone Insights app. Reported by @peterdewitte 
- Asset page doesn't break anymore where the URL parameters were mismatching. Reported by @peterdewitte 
- Fix for Gauge widget only supporting attributes with datapoints.
- Improvements for Appearance page file handling; no crashes anymore during local development.
Fixes #1093

pankalog committed:
AssetStateDuration (session) widget implementation

richturner committed:
Added endpoint to upsert predicted datapoints (Closes #1129)

Added endpoint to upsert predicted datapoints (Closes #1129)

DonWillems committed:
Enhancement/apply or-translate (#1218)

- Fixed some translations that came up with the introduction of or-translate to or-mwc-input type BUTTON
- Removed i18next from uses of the or-mwc-input type button
-Added or-translate to snackbar and removed i18next from snackbar instances
- Added or-translate to or-mwc-list so that e.g. the top right menu in the manager immediately translates when switching languages
-Applied or-translate to the settings drawers of widgets on the insights page

richturner committed:
CI/CD: Fix build issue (#1219)

richturner committed:
UI update: Fix typescript build issue (#1220)

pankalog committed:
Merge branch 'master' into feature/fleet-management

# Conflicts:
#	ui/component/or-dashboard-builder/src/index.ts

pankalog committed:
Merge remote-tracking branch 'origin/master' into feature/fleet-management

pankalog committed:
Update submodule

pankalog committed:
Merge remote-tracking branch 'origin/feature/fleet-management' into feature/fleet-management

# Conflicts:
#	ui/component/or-dashboard-builder/src/index.ts

pankalog committed:
Bug fixes

pankalog committed:
Use `AssetDatapointAllQuery` instead of LTTB for more accurate display of Teltonika datapoints

pankalog committed:
Better Reverse Geocode support and formatting
* New Nominatim client for better support
* Fixed formatting for Overview Analysis Type

pankalog committed:
Fix Analysis Types
Analysis types now more statically defined and easier to deal with. Functionality for multiple analysis types to come.

pankalog committed:
Update import

pankalog committed:
Rework Input/Output refs

Reworked to include the ref's attribute's type, used for determining what types of analysis should be available, and for casting/display purposes.

pankalog committed:
Stop double-rendering

Remove the connectedCallback() actions, as they were being triggered two times, once for the connectedCallback and once for firstUpdated. firstUpdated suffices, we just initialize everything later down the lifecycle of the LitElement

pankalog committed:
Improve UI and console logs

Remove a lot of console.logs and also updated UI element to explain the reasoning for the displayed error.

pankalog committed:
Attributes Panel Fix

This fix now properly updates the attributes within AttributeActionEvent:detail, so that the attribute included is the actual one that is selected

pankalog committed:
Fix console logs

pankalog committed:
LONG. TIME. COMING.

EVEN IF you handle the exception, it still prints it, taking up ENORMOUS amounts of space in my log. Just destroys the great work I've done over months of printing ONE LINE per correctly parsed payload. Very happy to have taken the executive decision for this. Took some emotional strength, but I finally am committing the removal of this messy-for-my-logs piece of code.

pankalog committed:
Merge branch 'master' into feature/fleet-management

pankalog committed:
Merge branch 'master' into feature/fleet-management

pankalog committed:
Merge remote-tracking branch 'origin/master' into feature/fleet-management

pankalog committed:
Merge branch 'master' into feature/fleet-management

pankalog committed:
Update to latest OpenRemote submodule

Theoretically this should fix everything?

pankalog committed:
Use the Keycloak official healthcheck instead of cURLing the homepage

richturner commented on model/src/main/java/org/openremote/model/util/ValueUtil.java:
```diff
@@ -216,7 +216,7 @@ public static <T> Optional<T> parse(String jsonString, Type type) {
             return Optional.of(JSON.readValue(jsonString, JSON.constructType(type)));
         } catch (Exception e) {
-            LOG.log(Level.WARNING, "Failed to parse JSON", e);
+//            LOG.log(Level.WARNING, "Failed to parse JSON", e);
```
Wouldn't merge this

richturner commented on ui/app/manager/src/pages/page-map.ts:
```diff
@@ -341,33 +370,66 @@ export class PageMap extends Page<MapStateKeyed> {
-                    return b.attributes[WellknownAttributes.LOCATION].value.coordinates[1] - a.attributes[WellknownAttributes.LOCATION].value.coordinates[1];
-                } else {
-                    return;
+                        this._assets.filter((asset) => {
+                            if (!asset.attributes) {
+                                return false;
+                            }
+                            const attr = asset.attributes[WellknownAttributes.LOCATION] as Attribute<GeoJSONPoint>;
+                            return !attr.meta || !attr.meta.hasOwnProperty(WellknownMetaItems.SHOWONDASHBOARD) || !!Util.getMetaValue(WellknownMetaItems.SHOWONDASHBOARD, attr);
+                        })
+                                .sort((a, b) => {
+                                    if (a.attributes[WellknownAttributes.LOCATION].value && b.attributes[WellknownAttributes.LOCATION].value) {
+                                        return b.attributes[WellknownAttributes.LOCATION].value.coordinates[1] - a.attributes[WellknownAttributes.LOCATION].value.coordinates[1];
+                                    } else {
+                                        return;
+                                    }
+                                })
+                                .map(asset => {
+                                    return html`
+                                        <or-map-marker-asset
+                                                ?active="${this._currentAsset && this._currentAsset.id === asset.id}"
+                                                .asset="${asset}"
+                                                .config="${this.config.markers}"></or-map-marker-asset>
+                                    `;
+                                })
```
Reformatting here makes it difficult to follow what has actually changed...we do need to enforce some linting and coding standards though

richturner commented on ui/app/manager/src/pages/page-map.ts:
```diff
@@ -341,33 +370,66 @@ export class PageMap extends Page<MapStateKeyed> {
+                <or-map-location-history-overlay .assetId="${this._currentAsset?.id}" .map="${this._map}" ></or-map-location-history-overlay>
+${currentAssetSelected ? html`
+                <or-map-asset-card .config="${this.config?.card}" .assetId="${this._currentAsset.id}"
+                                   .markerconfig="${this.config?.markers}" .map="${this._map}"></or-map-asset-card>
```
This newly added map property doesn't appear to be used

richturner commented on ui/app/manager/src/pages/page-map.ts:
```diff
@@ -341,33 +370,66 @@ export class PageMap extends Page<MapStateKeyed> {
+<!--                
+                <or-map-location-history-markers .assetId="${this._currentAsset?.id}" .map="${this._map}"></or-map-location-history-markers>
+            -->
+                <or-map-location-history-overlay .assetId="${this._currentAsset?.id}" .map="${this._map}" ></or-map-location-history-overlay>
```
If this location history is to be integrated into the generic manager app then it will need to be configurable

richturner commented on ui/component/or-attribute-card/src/index.ts:
```diff
@@ -744,11 +744,11 @@ export class OrAttributeCard extends LitElement {
             {
-                type: "lttb",
+                type: "all",
                 fromTimestamp: this._startOfPeriod,
                 toTimestamp: this._endOfPeriod,
                 amountOfPoints: 20,
-            } as AssetDatapointLTTBQuery
+            } as AssetDatapointAllQuery
```
Is this intentional?

MartinaeyNL replied:
To additionally comment on this; we're intentionally using `LTTB` in the `or-attribute-card` here.

richturner replied:
Thanks but is it a change that makes sense for everything or just this fleet management application?

richturner commented on ui/component/or-dashboard-builder/src/panels/attributes-panel.ts:
```diff
@@ -193,7 +193,7 @@ export class AttributesPanel extends LitElement {
 
     protected async loadAssets(): Promise<Asset[]> {
-        if(this.attributeRefs.filter(ar => !this.getLoadedAsset(ar)).length > 0) {
+        if(true) {
```
Remove or fix

richturner commented on ui/component/or-dashboard-builder/src/util/nominatim-browser.ts:
I think we can just update the forward/reverse geocoder functions in mapwidget rather than this, as it doesn't add enough functionality to warrant the copy paste and source code isn't maintained. We should definitely default to nominatim openstreetmap API though.

richturner commented on ui/component/or-dashboard-builder/src/widgets/session-widget.ts:
Need a way of being able to plugin custom widgets at runtime from a custom project

richturner commented on ui/component/or-map/src/mapwidget.ts:
```diff
@@ -873,6 +911,15 @@ export class MapWidget {
         }
     }
 
+    public async getGeoJsonFeature(query: {query: string}){
+        // let gc: MaplibreGeocoder. = this._geocoder;
+    }
```
Does this not work? Can it be fixed or removed or at least commented out with reason

richturner commented on ui/component/or-map/src/mapwidget.ts:
```diff
@@ -873,6 +911,15 @@ export class MapWidget {
+    public async getGeoJsonFeature(query: {query: string}){
+        // let gc: MaplibreGeocoder. = this._geocoder;
+    }
+    public async getAddress(query: {lat: number, lon: number}){
+        // let gc: MaplibreGeocoder. = this._geocoder;
+        return this._reverseGeocode(query);
+    }
+
+
```
Maybe should be exposed from the `or-map` so no dependency on the map for geocoding

richturner commented on ui/component/or-map/src/markers/or-map-asset-card-trip-section.ts:
This needs to be a pluggable mechanism from a custom project

richturner commented on ui/component/or-map/src/or-map-location-history-markers.ts:
```diff
@@ -0,0 +1,112 @@
+            // el.className = "historyMarker"
+
+            const time: string = moment(historyElement.x).toString();
+            const popup = new maplibregl.Popup().setText(
+                'Construction on the Washington Monument began in 1848.'
```
Cleanup required

richturner commented on ui/component/or-map/src/or-map-location-history-markers.ts:
I like this as a marker type but needs to be configurable at the map level.

Can this marker not extend the OrMapMarkerAsset or even just add this functionality to that marker.

History duration to retrieve in map config (if set then uses this marker instead of asset marker)

richturner commented on ui/component/or-map/src/orMapLocationHistoryOverlay.ts:
Not clear to me why this is needed rather than encapsulating everything in the history marker

pankalog committed:
Merge branch 'refs/heads/master' into feature/fleet-management

# Conflicts:
#	container/src/main/java/org/openremote/container/security/keycloak/KeycloakIdentityProvider.java
#	ui/component/or-dashboard-builder/src/index.ts
#	yarn.lock

pankalog committed:
Merge branch 'refs/heads/master' into feature/fleet-management

pankalog committed:
Fix UI errors during build (Interface changes for GeoJSONSource)

pankalog committed:
Register widget with or-dashboard-builder

pankalog committed:
Merge branch 'refs/heads/master' into feature/fleet-management

pankalog committed:
Merge branch 'refs/heads/master' into feature/fleet-management

pankalog committed:
Merge branch 'master' into feature/fleet-management

# Conflicts:
#	ui/component/or-attribute-card/src/index.ts
#	yarn.lock

pankalog committed:
Merge remote-tracking branch 'origin/master' into feature/fleet-management

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/openremote/openremote?pullRequest=1147) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/openremote/openremote?pullRequest=1147) before we can accept your contribution.<br/>**0** out of **5** committers have signed the CLA.<br/><br/>:x: peterdewitte<br/>:x: MartinaeyNL<br/>:x: robin-dekkers<br/>:x: Miggets7<br/>:x: DonWillems<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/openremote/openremote?pullRequest=1147) it.</sub>
